tempesta_fw/classifier/frang.c

```diff
@@ -0,0 +1,905 @@
+	ptr = buf + sizeof("Host:") - 1;
+	ptr = skip_spaces(ptr);
+	if (!tfw_addr_pton(ptr, &addr))
+		return TFW_BLOCK;
+	return TFW_PASS;
+}
+
+enum {
+	Frang_Req_0,
+
+	Frang_Req_Hdr_Start,
+	Frang_Req_Hdr_Method,
+	Frang_Req_Hdr_UriLen,
+	Frang_Req_Hdr_FieldDup,
+	Frang_Req_Hdr_FieldLenRaw,
+	Frang_Req_Hdr_FieldLenSpecial,
+	Frang_Req_Hdr_Crlf,
+	Frang_Req_Hdr_Host,
+	Frang_Req_Hdr_ContentType,
+
+	Frang_Req_Hdr_NoState,
+
+	Frang_Req_Body_Start,
+	Frang_Req_Body_Timeout,
+	Frang_Req_Body_Len,
+
+	Frang_Req_Body_NoState,
+
+	Frang_Req_NothingToDo
+};
+
+#define FSM_HDR_STATE(state)						\
+	((state > Frang_Req_Hdr_Start) && (state < Frang_Req_Hdr_NoState))
+
+#define __FSM_INIT()							\
+int __fsm_const_state;
+
+#define __FSM_START(st)							\
+switch(st)
+
+#define __FSM_FINISH()							\
+done:									\
+	TFW_DBG("Finish FRANG FSM at state %d\n", __fsm_const_state);	\
+	TFW_DBG("Return %s\n", r == TFW_PASS ? "PASS" : "BLOCK");	\
+	req->frang_st = __fsm_const_state;
+
+#define __FSM_STATE(st)							\
+case st:								\
+st: __attribute__((unused))						\
+	TFW_DBG("enter FRANG FSM at state %d\n", st);			\
+	__fsm_const_state = st; /* optimized out to constant */
+
+#define __FSM_EXIT()	goto done;
+
+#define __FSM_JUMP(to)	goto to;
+#define __FSM_MOVE(to)							\
+	if (r)								\
+		__FSM_EXIT();						\
+	goto to;
+
+#define __FSM_JUMP_EXIT(to)						\
+	__fsm_const_state = to; /* optimized out to constant */		\
+	__FSM_EXIT()
+
+static int
+frang_http_req_handler(void *obj, unsigned char *data, size_t len)
+{
+	int r = TFW_PASS;
+	unsigned int body_len = len;
+	TfwConnection *conn = (TfwConnection *)obj;
+	TfwHttpReq *req = container_of(conn->msg, TfwHttpReq, msg);
+	struct sk_buff *head_skb = (void *)ss_skb_peek(&req->msg.skb_list);
+	struct sk_buff *skb = (void *)ss_skb_peek_tail(&req->msg.skb_list);
+
+	__FSM_INIT();
+
+	/*
+	 * There's no need to check for header timeout if this is the very
+	 * first data chunk of a request (first full separate SKB with data.
+	 * The FSM is guaranteed to go through the initial states and then
+	 * either block or move to one of header states. Then header timeout
+	 * is checked on each consecutive SKB with data.
+	 *
+	 * Why is this not one of FSM states? Basically, that's to avoid
+	 * going through unnecessary FSM states each time this is run. When
+	 * there's a slowris attack, we may stay long in Hdr_Method or in
+	 * Hdr_UriLen states, and that would require including the header
+	 * timeout state in the loop. But when we're past these states, we
+	 * don't want to run through them on each run again, and just loop
+	 * in FieldDup and FieldLen states. I guess that can be done with
+	 * some clever FSM programming, but this is plain simpler.
+	 */
+	if (frang_cfg.clnt_hdr_timeout
+	    && (skb != head_skb) && FSM_HDR_STATE(req->frang_st)) {
+		unsigned long start = req->tm_header;
+		unsigned long delta = frang_cfg.clnt_hdr_timeout;
+		if (time_is_after_jiffies(start + delta))
```

Comment:
We should do TFW_WARN() on reaching limits for this module. Please correct this and check other places.
Currently we just log the messages in dmesg and surely this is bottleneck in DDoS. The more performant logger is in our TODO.